internal/cli/cmd/cluster/artifact.go

```diff
@@ -102,24 +125,160 @@ func newArtifactDownloadCmd() *cobra.Command {
+				// Fallthrough.
+			}
 		}
 
-		fmt.Fprintf(console.Stdout(ctx), "Downloaded %s (namespace %s) to %s", src, namespace, dest)
+		cachePath := cacheArtifactPath(time.Now(), parsedURL)
+		labels := map[string]string{cacheSourceURLLabel: sourceURL}
+
+		backoff.RetryNotify(func() error {
+			req, err := http.NewRequestWithContext(ctx, "GET", sourceURL, nil)
+			if err != nil {
+				return backoff.Permanent(fnerrors.InvocationError("remote url", "failed to prepare request: %w", err))
+			}
+
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				return fnerrors.InvocationError("remote url", "failed to send request: %w", err)
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode >= http.StatusInternalServerError {
+				return fnerrors.InvocationError("remote url", "remote server returned status code %d", resp.StatusCode)
+			} else if resp.StatusCode != http.StatusOK { // treat non-server error as permanent
+				return backoff.Permanent(fnerrors.InvocationError("remote url", "remote server returned status code %d", resp.StatusCode))
+			}
 
+			w, err := os.Create(dest)
+			if err != nil {
+				return backoff.Permanent(fnerrors.New("failed to open %q: %w", dest, err))
+			}
+			defer w.Close()
+
+			teeR := io.TeeReader(resp.Body, w)
+
+			var length int64
+			if cl := resp.Header.Get("content-length"); cl != "" {
+				if n, err := strconv.ParseInt(cl, 10, 64); err != nil {
+					return fnerrors.InvocationError("remote url", "remote server returned invalid content-length (%s): %w", cl, err)
+				} else {
+					length = n
+				}
+			}
+
+			if err := storage.UploadArtifactWithOpts(ctx, cli, cacheArtifactNamespace, cachePath, teeR, storage.UploadOpts{
+				Labels: labels,
+				Length: length,
+			}); err != nil {
+				return fnerrors.InvocationError("remote url", "failed to cache artifact: %w", err)
+			}
+
+			return nil
+		},
+			backoff.WithMaxRetries(backoff.WithContext(backoff.NewConstantBackOff(5*time.Second), ctx), cacheURLRetries),
+			func(err error, delay time.Duration) {
+				fmt.Fprintf(console.Stderr(ctx), "Error: Failed to cache artifact: %v; retrying in %v...\n", err, delay)
+			})
+
+		fmt.Fprintf(console.Stderr(ctx), "Downloaded to %s\n", dest)
 		return nil
 	})
 }
+
+func writeArtifact(ctx context.Context, cli storage.Client, namespace, path string, dest string) error {
+	reader, err := storage.ResolveArtifactStream(ctx, cli, namespace, path)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	w, err := os.Create(dest)
+	if err != nil {
+		return fnerrors.New("failed to open %q: %w", dest, err)
+	}
+	defer w.Close()
+
+	if _, err := io.Copy(w, reader); err != nil {
+		return fnerrors.New("failed to write %q: %w", dest, err)
+	}
+	return nil
+}
+
+func cacheArtifactPath(now time.Time, sourceURL *url.URL) string {
+	safePath := slashRuns.ReplaceAllString(strings.Join(validPath.FindAllString(sourceURL.Path, -1), "-"), "/")
```

Comment:
FindAllString feels wrong here. Wouldn't this match the full path?

Reply:
It will match all safe parts of the path. Then we'll merge them. In principle we just need to get rid of unsafe characters.